set_test.go

```diff
@@ -147,3 +235,105 @@ func TestSetIter(t *testing.T) {
 		t.Fatalf("unexpected sum: %d, want %d", sum, sumTo99)
 	}
 }
+
+func testSetMemoryUsage[K IntKey]() func(t *testing.T) {
+	return func(t *testing.T) {
+		s := NewSet[K](16)
+		m := New[K, struct{}](16)
+		szs := int(unsafe.Sizeof(s.data[0]))
+		msz := int(unsafe.Sizeof(m.data[0]))
+		if szs*2 < msz {
+			t.Fatalf("The per-element size of Set[%T] (%d) should be half that of Map[%T] (%d)",
+				s.data[0], szs, m.data[0], msz)
+		}
+		t.Logf("Sizeof(Set[%T]) = %d; Sizeof(Map[%T]) = %d",
+			s.data[0], szs, m.data[0], msz)
+	}
+}
+
+func TestSetMemoryUsage(t *testing.T) {
+	t.Run("int", testSetMemoryUsage[int]())
+	t.Run("int8", testSetMemoryUsage[int8]())
+	t.Run("int16", testSetMemoryUsage[int16]())
+	t.Run("int32", testSetMemoryUsage[int32]())
+	t.Run("int64", testSetMemoryUsage[int64]())
+
+	t.Run("uint", testSetMemoryUsage[uint]())
+	t.Run("uint8", testSetMemoryUsage[uint8]())
+	t.Run("uint16", testSetMemoryUsage[uint16]())
+	t.Run("uint32", testSetMemoryUsage[uint32]())
+	t.Run("uint64", testSetMemoryUsage[uint64]())
+}
+
+func BenchmarkSetAdd10K(b *testing.B) {
+	rr := rand.New(rand.NewSource(12345))
+	seen := make(map[int]struct{}, 10_000)
+	for len(seen) < 10_000 {
+		seen[int(rr.Int31())] = struct{}{}
+	}
+	keys := slices.AppendSeq(make([]int, 0, 10_000), maps.Keys(seen))
+
+	s := NewSet[int](10_000)
+	b.ResetTimer()
+
+	j := 0
+	for i := 0; i < b.N; i++ {
+		s.Add(keys[j])
+		j++
+		if j == 10_000 {
+			j = 0
+			s.Clear()
+		}
+	}
+}
+
+func BenchmarkSetDel10K(b *testing.B) {
+	rr := rand.New(rand.NewSource(12345))
+
+	s := NewSet[int](10_000)
+	keys := make([]int, 0, 10_000)
+	for s.Len() < 10_000 {
+		v := rr.Int()
+		if s.Add(v) {
+			keys = append(keys, v)
+		}
+	}
+
+	// Duplicate S to make resetting it back
+	// to a populated state faster.
+	dup := *s
+	dup.data = slices.Clone(dup.data)
+
+	b.ResetTimer()
+
+	j := 0
+	for i := 0; i < b.N; i++ {
+		s.Del(keys[j])
+		j++
+		if j == 10_000 {
+			j = 0
+			b.StopTimer()
+			copy(s.data, dup.data)
+			s.size = dup.size
+			s.hasZeroKey = dup.hasZeroKey
+			b.StartTimer()
+		}
+	}
+}
+
+func BenchmarkSetHas(b *testing.B) {
+	rr := rand.New(rand.NewSource(12345))
+	s := NewSet[int32](256)
+	var keys [32]int32
+	for i := 0; s.Len() < 256; i++ {
+		v := int32(rr.Int31())
+		if s.Add(v) && i < len(keys) {
+			keys[i] = v
+			i++
+		}
+	}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		s.Has(keys[i%len(keys)])
+	}
```

Comment:
This is a bit hard to follow, and again uses the benchmark index for probing. Can be simplified to

```suggestion
	keys := make([]int32, 256)
	for i := range 256 {
		v := rr.Int31()
		keys[i] = v
		s.Add(v)
	}

	b.ResetTimer()
	for range b.N {
		for _, k := range keys {
			_ = s.Has(k)
		}
	}
```

Reply:
Question: would you like to change all the `Set` benchmarks to measure the time it takes to process all of the keys (your suggestions) or the time it takes to measure a single `{Add,Del,Has}` op (current code)? I don't have a strong opinion here and am happy to make the change - so let me know.

Also, thank you again for taking the time to review this PR!

Reply:
All good :-D

I normally prefer the benchmark loop body to contain a non-trivial chunk of work. I have found that this produces more stable benchmark results. So not just a single invoke of x.Foo(y), but an extra loop exercising x.Foo several times.

I also prefer to have each iteration of the loop execute the _exact_ same thing. If the benchmark loop index is used to pass into Add,Del,Has there could theoretically be some subtle alignment bias when b.N is varied (I don't think there will be any significant bias for your particular code, but generally speaking).